Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -145,8 +162,33 @@ public struct HTTPResponseHead: Equatable {
     }
 }
 
-fileprivate typealias HTTPHeadersStorage = [String: [(String, String)]] // [lowerCasedName: [(originalCaseName, value)]
+/// The Index for a header name or value that points into the underlying `ByteBuffer`.
+struct HTTPHeaderIndex {
+    let start: Int
+    let length: Int
+}
+
+struct HTTPHeader {
+    let name: HTTPHeaderIndex
+    let value: HTTPHeaderIndex
+}
 
+private extension ByteBuffer {
+
+    func equalCaseInsensitiveASCII<S: Sequence>(seq: S, at: HTTPHeaderIndex) -> Bool where S.Element == UInt8 {
+        return withVeryUnsafeBytes { buffer in
+            let address = buffer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+            var idx = 0
+            for byte in seq {
+                guard address.advanced(by: at.start + idx).pointee & 0xdf == byte & 0xdf else {
```

Comment:
This might slow us down slightly, but do we want to also enforce that both bytes are actually in the ASCII range? This function will misbehave if it's handed non-ASCII bytes (e.g. ISO-8859-1).